### DISCUSSION TITLE
Very long sleep possible?

### DISCUSSION THREAD

**Discussion by cpainchaud** [SpenceKonde/megaTinyCore#413] [Q&A] [answered]
Hi,

 So I went through sleep modes pages and code examples but I feel like I  am missing a piece to make the device sleep for  long time (like 1+ hours). Using the 1K clock and biggest period RTC_PERIOD_CYC32768_gc , I can achieve like 32second maximum. 
 
  What is the obvious thing I am currently missing ?

thank you !

**Comment by freemovers** [comment, marked as answer]
Did you change the prescaler as well? Maximum prescaler is RTC_PRESCALER_DIV32768_gc, but looks like yours is set to RTC_PRESCALER_DIV1_gc.

**Reply by cpainchaud** [reply]
Hi, I was looking into prescalers but I was afraid to break anything else as there's not much documentation on these so I don't knwo the consequences of changing them.

Do you have any pointers/code on how to use them properly?

**Reply by freemovers** [reply]
I'm not able to test this right now, but the following code should work. The pre-scaler is specific for the RTC, so it will not affect other peripherals, not even the PIT.

```c
void RTC_init(void)
{
	RTC.CLKSEL = RTC_CLKSEL_INT1K_gc;         // 1kHz Internal Crystal Oscillator
	while (RTC.STATUS > 0 || RTC.PITSTATUS);  // Wait for all register to be synchronized
	RTC.PER = 0x70;                           // Max for overflow of 112 (112*32 = 3584 ~ 1 hour)
	RTC.INTCTRL |= RTC_OVF_bm;                // Enable overflow Interrupt which will trigger ISR
	RTC.CTRLA = RTC_PRESCALER_DIV32768_gc     // 32768 / 1000 = 32 sec.
	| RTC_RTCEN_bm                            // Enable: enabled
	| RTC_RUNSTDBY_bm;                        // Run In Standby: enabled
}
```

**Reply by cpainchaud** [reply]
I am trying to adapt to smaller scale (1 minute) so test your code but it never exists from sleep.

it could be because I am using the wrong ISR() and stuff (INTCTRL and CTRLA are not PITCTRLA for example ....)

**Reply by cpainchaud** [reply]
````
void RTC_init()
{
  /* Initialize RTC: */
  while (RTC.STATUS > 0 || RTC.PITSTATUS)
  {
    ;
  }
  RTC.CLKSEL = RTC_CLKSEL_INT1K_gc;   

  RTC.INTCTRL |= RTC_OVF_bm;
  /* PIT Interrupt: enabled */

  RTC.CTRLA = RTC_PRESCALER_DIV32768_gc     // 32768 / 1000 = 32 sec.
              | RTC_RTCEN_bm                            // Enable: enabled
              | RTC_RUNSTDBY_bm;                        // Run In Standby: enabled

   RTC.PER = 0x2;
   
}

ISR(RTC_PIT_vect)
{
  RTC.PITINTFLAGS = RTC_PI_bm;          /* Clear interrupt flag by writing '1' (required) */
}

````

**Reply by freemovers** [reply]
PIT is different from the RTC. They are using the same clock source, but that is basically it. In your code you are setting up the RTC, but you don't have the ISR for the RTC. You need the following:
```C
ISR(RTC_CNT_vect)
{
       RTC.INTFLAGS = RTC_OVF_bm;				// Clear flag by writing '1':
}
```
Unfortunately the RTC does not work in sleep mode power down. The PIT does work in sleep mode power down, but the PIT *does not* use the prescaler. So with the PIT you are limited to 32 seconds.

**Reply by cpainchaud** [reply]
Hi!

This is making sense, I tried ISR(RTC_CNT_vect) but it didnt work and you just exposed the reason.

It's ironic .... DEEP SLEEP cannot be achieved with long timer .....

Thank you for your help anyway !

**Reply by grandaspanna** [reply]
Have a look at the code I posted the other day in another discussion in this repository. In standby mode, using the RTC interrupt, I was measuring single digit microamps so you can use the RTC in that mode. By setting the prescaler and appropriate target, the datasheet says you can get to 18 hours. If you have to use the PIT, then maybe just keep a counter and increment it each time the interrupt is generated if the interval is too short. You can scale up as big as you want this way. It means a short piece of code running each interrupt which will raise power consumption (someone posted that they got overall lower power usage with a higher clock rate because it reduced the runtime of the code, but YMMV).

**Reply by cpainchaud** [reply]
thank  @grandaspanna I came to the same conlusion!

**Reply by amadeuspzs** [reply]
@grandaspanna do you have an example of how to implement:

> If you have to use the PIT, then maybe just keep a counter and increment it each time the interrupt is generated if the interval is too short.

?

I have the PIT setup to wake every 8s, and want to set a total sleep in increments of this:

```c
volatile int sleeps = 0; // how many sleeps, used in conjuction with:
int TRANSMITPERIOD = 16; //transmit a packet to gateway so often (in s), factor of 8
```

```c
ISR(RTC_PIT_vect)
{
  sleeps += 1; // will this work as PIT is a reset, ie it will lose state?!
  RTC.PITINTFLAGS = RTC_PI_bm;          /* Clear interrupt flag by writing '1' (required) */
}
```

```c
void loop() {
  if (sleeps == 0 || sleeps >= (TRANSMITPERIOD / 8) ) {
    // transmit here
  }
  sleep_cpu();
}
```

The PIT interrupt triggers a reset, so the `sleep` variable will lose it's value?

Currently, it's just waking up every 8 seconds?

I'd like to get to 10 minute wake up cycles ultimately (8s is just for testing!).

**Reply by grandaspanna** [reply]
I wouldn't have thought that the wakeup was generating a reset - well, that's not what I'd expect.

What does your RTC setup look like?

**Reply by amadeuspzs** [reply]
You are correct; I was confusing the wakeup behaviour with the ESP8266 which is a reset.

It's all working nicely: https://github.com/amadeuspzs/attiny1614-rtc/blob/main/attiny1614-rtc-blink-long-sleep/attiny1614-rtc-blink-long-sleep.ino

Thanks for the help!

**Comment by SpenceKonde** [comment]
Thanks for the code examples. A future version of the core which addresses the longstanding issue https://github.com/SpenceKonde/megaTinyCore/issues/158 will add a library tentatively called SleepLib. which would have the following features: 
* Switch to RTC millis and sleep. If woken by the RTC overflow interrupt, adjust current millis count and go back to sleep. If woken for other reasons, adjust millis based on current RTC count, stop the RTC, and switch back to the normal millis timer and resume user code
* Switch to RTC millis. Wake every after a configurable period of time and switch back to a HF timer. 
* Handle the incredibly nasty errata on most 0/1-series relating to the RTC. This bug is described in the errata as if it were a ladybug, when it';s closer to that bad guy in Men in Black, and more confusing too. Also the only errata on modern AVRs that will be impossible to fix without infuriating some people, because there is a way to exploit it to get otherwise impossible functionality. 
* (maybe)  An option to wake every (period of time) call a user defined function, and then go back to sleep (to make dataloggers supereasy). Same RTC treatment as above. 
* (maybe) debug option: Count number of times there is a wake determined to be just related to RTC which never runs user code, and number of times there's a wake that runs user code. 
* .(hopefully) if library is included, find way to warn people or throw an error if they have left in something that goes to sleep while bypassing the new methods (ie, ensure that sleep_cpu() at minimum gives a warning, and preferably won't compile. 
* (tentative due to hardware bugs) Try to make USART SOF interrupt work (may be extremely challenging, see errata) 
* Use Standby sleep so that user can choose to leave whichever peripherals they desire on to wake with, (in stark contrast with classic AVRs where standby sleep barely saves any power whatsoever, if you don't have anything's RUNSTBY bit set, the power consumption on tinyAVR is supposed to be the same as power down! This may only be the case on tinyAVR - in which case the Dx and such wuiold need a way to use powerdown.
* Add a more convenient facility to disable digital input buffers in order to save power - it is widely known that despite insistance to the contrary from some corners, floating pins are power hogs on modern AVRs.. (#695 )
* Support all modern AVRs with a single library.

**Comment by SpenceKonde** [comment]
What the heck is this sorcery? I got an email about a reply here, with a screenshot of a figure from the datasheet that I'm kind of confused by. The words in the datasheet imply that the prescaled clock is what gets counted by the PIT, but I'm not sure how that timing diagram doesn't show the opposite, which would make it inaccurate.

Example 23-1. PIT Timing Diagram for PERIOD = CYC16

This diagram unambiguously shows the prescaled clock going to the PIT

Figure 23-2. Timing Between PIT Enable and First Interrupt

But this one says that the PIT that the PIT is derived from the state of the RTC prescaler, not something attached 

Do note also that the ATtiny 0-series and 1-series (except with rev C 1-series parts with 32k of flash), have some very mean chameleon hornets nesting behind the CTRLA registers, try to avoid using impacted parts if you need the RTC, and if you must, you must never turn on the RTC without also turning on the PIT before any further interaction with the peripheral. As long as you do that, you'll be fine, as long as you don't touch RTC.CTRLA. If you do, the hornets still get one chance to sting you every time you write to it, resetting the current RTC period.

When they finally fix that, it's gonna be a fuckin shitshow. I guarantee you there will be multiple instances of devices in mass production that didn't realize that was a bug, or even that something else wasn't resetting it (nothing is supposed to short of a reset or turning off the PIT), or completely misapprehended how the PIT works (I went back and forth with someone for days before it was made plain that they expected to be able to set the delay, go to sleep, and have the pit wake them that many second later. Uh, sorry, nope, not on the menu, not on those parts or any other AVRs. You can have the RTC wake you like that, but he wanted to use power down sleep mode. 

Thinking about it - it has to be that he didn't know that the only thing about standby sleep unchanged since classic AVRs is the name - it used to be something a little lower power than IDLE, but only a little, because the main power hog at that point is the oscillator - which it explicitly keeps on to eliminate the startup time of an external crystal; that was it's only purpose. Now it's an omnibus deeper-than-idle-less-deep-than-power-down mode, which can be tuned to anything from being lit up like a christmas tree, sucking down near-IDLE current, or you can leave everything off and it's nearly identical to power-down.
